dirty_cat/_fuzzy_join.py

```diff
@@ -7,31 +7,155 @@
+
+    Parameters
+    ----------
+    main : pd.DataFrame
+        A table with string columns.
+    main_cols : str or list
+        The columns of the main table.
+    aux : pd.DataFrame
+        Another table with string columns.
+    aux_cols : str or list
+        The columns of the aux table.
+    encoder: Union[Literal["hashing"], _VectorizerMixin]
+        Encoder parameter for the Vectorizer.
+        See fuzzy_join's docstring for more information.
+    analyzer : {"word", "char", "char_wb"}
+        Analyzer parameter for the HashingVectorizer passed to
+        the encoder and used for the string similarities.
+        See fuzzy_join's docstring for more information.
+    ngram_range : int 2-tuple, optional, default=(2, 4)
+        The lower and upper boundary of the range of n-values for different
+        n-grams used in the string similarity.
+        See fuzzy_join's docstring for more information.
+
+    Returns
+    -------
+    main_array : array-like
+        An array of the encoded columns of the main table.
+    aux_array : array-like
+        An array of the encoded columns of the aux table.
+    """
+    # Make sure that the column types are string and categorical:
+    main_cols_clean = main[main_cols].astype(str)
+    aux_cols_clean = aux[aux_cols].astype(str)
+
+    if isinstance(main_cols, list) and isinstance(aux_cols, list):
+        main_cols_clean = main_cols_clean[main_cols].apply(
+            lambda row: "  ".join(row.values.astype(str)), axis=1
+        )
+        aux_cols_clean = aux_cols_clean[aux_cols].apply(
+            lambda row: "  ".join(row.values.astype(str)), axis=1
+        )
+    all_cats = pd.concat([main_cols_clean, aux_cols_clean], axis=0).unique()
+
+    if encoder is None:
+        enc = HashingVectorizer(analyzer=analyzer, ngram_range=ngram_range)
+    else:
+        enc = encoder
+
+    enc_cv = enc.fit(all_cats)
+    main_enc = enc_cv.transform(main_cols_clean)
+    aux_enc = enc_cv.transform(aux_cols_clean)
+
+    all_enc = vstack((main_enc, aux_enc))
+
+    tfidf = TfidfTransformer().fit(all_enc)
+    main_enc = tfidf.transform(main_enc)
+    aux_enc = tfidf.transform(aux_enc)
+    return main_enc, aux_enc
+
+
+def _nearest_matches(main_array, aux_array):
+    """Find the closest matches using the nearest neighbors method.
+
+    Parameters
+    ----------
+    main_array : array-like
+        An array of the encoded columns of the main table.
+    aux_array : array-like
+        An array of the encoded columns of the aux table.
+
+    Returns
+    -------
+    idx_closest
+        Index of the closest matches of the main table in the aux table.
+    norm_distance
+        Distance between the closest matches, on a scale between 0 and 1.
+    """
+    # Find nearest neighbor using KNN :
+    neigh = NearestNeighbors(n_neighbors=1)
+    neigh.fit(aux_array)
+    distance, neighbors = neigh.kneighbors(main_array, return_distance=True)
+    idx_closest = np.ravel(neighbors)
```

Comment:
Since we're flattening `idx_closest`, it may be more intuitive to do the same with `distance`, WDYT?